File: sp/src/main/java/wut/sp/Models/Book.java
package wut.sp.Models;

import java.util.ArrayList;
import java.util.List;

import wut.sp.Interfaces.IElement;

public class Book extends Section{
    private String title;
    private List<Author> authors;
    private List<IElement> sections;

    public Book() {
    }

    public Book(String title) {
        this.title = title;
        this.authors = new ArrayList<Author>();
        this.sections = new ArrayList<IElement>();
    }

    public Book(String title, Author author) {
        this.title = title;
        this.authors = new ArrayList<Author>();
        this.authors.add(author);
        this.sections = new ArrayList<IElement>();
    }

    public void addAuthor(Author author) {
        this.authors.add(author);
    }

    public void addContent(IElement element) {
        this.sections.add(element);
    }

    public String getTitle() {
        return this.title;
    }

    public void print() {
        System.out.println("Book: " + this.title);
        System.out.println("\nAuthors:");
        for (Author author : this.authors) {
            author.print();
        }
        System.out.println();
        for (IElement section : this.sections) {
            section.print();
        }
    }
}

File: spC++/src/Implement/Models/Book.cpp
#include "Headers/Models/Book.hpp"

Book::Book(std::string title) : Section(title) {}

Book::Book(Book* book) : Section(book->title) {
    this->authorList = book->authorList;
}

void Book::addAuthor(Author author) {
    this->authorList.push_back(author);
}

void Book::accept(IVisitor* v) {
    v->visitBook(this);
}

std::list<Author> Book::getAuthors() {
    return this->authorList;
}